c/rce.h

```diff
@@ -0,0 +1,382 @@
+                                       &cell_data, index);
+  CHECK(err);
+
+  RCDataType rc_data = make_RCData(&cell_data);
+
+  uint32_t item_id = rc_data.t->item_id(&rc_data);
+  if (item_id == RCDataUnionRule) {
+    RCRuleType rule = rc_data.t->as_RCRule(&rc_data);
+    // "Any more RCRule structures will result in an immediate failure."
+    CHECK2(rce_state->rcrules_count < MAX_RCRULES_COUNT,
+           ERROR_TOO_MANY_RCRULES);
+
+    rce_state->rcrules[rce_state->rcrules_count].flags = rule.t->flags(&rule);
+    mol2_cursor_t smt_root = rule.t->smt_root(&rule);
+    mol2_read_at(&smt_root,
+                 rce_state->rcrules[rce_state->rcrules_count].smt_root,
+                 SMT_KEY_BYTES);
+
+    rce_state->rcrules_count++;
+  } else if (item_id == RCDataUnionCellVec) {
+    RCCellVecType cell_vec = rc_data.t->as_RCCellVec(&rc_data);
+
+    uint32_t len = cell_vec.t->len(&cell_vec);
+    for (uint32_t i = 0; i < len; i++) {
+      uint8_t hash[BLAKE2B_BLOCK_SIZE];
+
+      bool existing = false;
+      mol2_cursor_t item = cell_vec.t->get(&cell_vec, i, &existing);
+      CHECK2(existing, ERROR_INVALID_MOL_FORMAT);
+      CHECK2(item.size == BLAKE2B_BLOCK_SIZE, ERROR_INVALID_MOL_FORMAT);
+
+      uint32_t read_len = mol2_read_at(&item, hash, sizeof(hash));
+      CHECK2(read_len == sizeof(hash), ERROR_INVALID_MOL_FORMAT);
+      err = rce_gather_rcrules_recursively(rce_state, hash, depth + 1);
+      CHECK(err);
+    }
+  } else {
+    CHECK2(false, ERROR_INVALID_MOL_FORMAT);
+  }
+
+  err = 0;
+exit:
+  return err;
+}
+
+int rce_collect_hashes(smt_state_t* bl_states, smt_state_t* wl_states) {
+  int err = 0;
+  uint32_t index = 0;
+
+  uint8_t lock_script_hash[SMT_KEY_BYTES];
+  uint64_t lock_script_hash_len = SMT_KEY_BYTES;
+
+  index = 0;
+  while (true) {
+    err = ckb_checked_load_cell_by_field(
+        lock_script_hash, &lock_script_hash_len, 0, index,
+        CKB_SOURCE_GROUP_INPUT, CKB_CELL_FIELD_LOCK_HASH);
+    if (err == CKB_INDEX_OUT_OF_BOUND) {
+      break;
+    }
+    err = smt_state_insert(wl_states, lock_script_hash, SMT_VALUE_EXISTING);
+    CHECK(err);
+    err = smt_state_insert(bl_states, lock_script_hash, SMT_VALUE_NOT_EXISTING);
+    CHECK(err);
+    index++;
+  }
+  index = 0;
+  while (true) {
+    err = ckb_checked_load_cell_by_field(
+        lock_script_hash, &lock_script_hash_len, 0, index,
+        CKB_SOURCE_GROUP_OUTPUT, CKB_CELL_FIELD_LOCK_HASH);
+    if (err == CKB_INDEX_OUT_OF_BOUND) {
+      break;
+    }
+    err = smt_state_insert(wl_states, lock_script_hash, SMT_VALUE_EXISTING);
+    CHECK(err);
+    err = smt_state_insert(bl_states, lock_script_hash, SMT_VALUE_NOT_EXISTING);
+    CHECK(err);
+    index++;
+  }
+
+  err = 0;
+exit:
+  return err;
+}
+
+int rce_validate(int is_owner_mode, size_t extension_index, const uint8_t* args,
+                 size_t args_len) {
+  int err = 0;
+  RceState rce_state;
+  rce_init_state(&rce_state);
```

Comment:
if `is_owner_mode`, this line will cost more cycles and needless. And this line is the same logic as `rce_state.rcrules_count = 0;` in line `296`. Suggestion: remove this line.

Reply:
already done.